src/Swarm/Language/Typecheck.hs

```diff
@@ -508,3 +539,117 @@ check t ty = do
   ty' <- infer t
   _ <- ty =:= ty'
   return ()
+
+-- | Ensure a term is a valid argument to @atomic@.  Valid arguments
+--   may not contain @def@, @let@, or lambda. Any variables which are
+--   referenced must have a primitive, first-order type such as
+--   @string@ or @int@ (in particular, no functions, @cmd@, or
+--   @delay@).  We simply assume that any locally bound variables are
+--   OK without checking their type: the only way to bind a variable
+--   locally is with a binder of the form @x <- c1; c2@, where @c1@ is
+--   some primitive command (since we can't refer to external
+--   variables of type @cmd a@).  If we wanted to do something more
+--   sophisticated with locally bound variables we would have to
+--   inline this analysis into typechecking proper, instead of having
+--   it be a separate, out-of-band check.
+--
+--   The goal is to ensure that any argument to @atomic@ is guaranteed
+--   to evaluate and execute in some small, finite amount of time, so
+--   that it's impossible to write a term which runs atomically for an
+--   indefinite amount of time and freezes the rest of the game.  Of
+--   course, nothing prevents one from writing a large amount of code
+--   inside an @atomic@ block; but we want the execution time to be
+--   linear in the size of the code.
+--
+--   We also ensure that the atomic block takes at most one tick,
+--   i.e. contains at most one tangible command. For example, @atomic
+--   {move; move}@ is invalid, since that would allow robots to move
+--   twice as fast as usual by doing both actions in one tick.
+validAtomic :: Syntax -> Infer ()
+validAtomic s@(Syntax l t) = do
+  n <- analyzeAtomic S.empty s
+  when (n > 1) $ throwError (InvalidAtomic l (TooManyTicks n) t)
+
+-- | Analyze an argument to @atomic@: ensure it contains no nested
+--   atomic blocks and no references to external variables, and count
+--   how many tangible commands it will execute.
+analyzeAtomic :: Set Var -> Syntax -> Infer Int
+analyzeAtomic locals (Syntax l t) = case t of
+  TUnit {} -> return 0
+  -- No nested 'atomic' allowed!
+  TConst Atomic -> throwError (InvalidAtomic l NestedAtomic t)
+  TConst c -> return $ if isTangible c then 1 else 0
```

Comment:
Oh, great catch! We should disallow `drill` too.

Reply:
Maybe it would be better to have a whitelist of allowed constants?  Otherwise I just know we are going to add a command in the future that takes multiple ticks and forget to check for it here.